### PR TITLE
Add `TorchTRTConversion` pass

### PR DESCRIPTION
**Pull request by jambayk**
## Describe your changes
This PR adds `TorchTRTConversion` pass which converts `nn.Linear` modules in transformer layers to `TRTModule` from `torch-tensorrt`. `torch-tensorrt` is an extension to `torch` where TensorRT engines can be used like normal `nn.Module`. 

We require this conversion to take advantage of `2:4` sparsity. 

~~I have not added unit test for the pass since the pass requires GPU and we don't have GPU support for ci tests yet. I will add a unit test in a follow up PR that uses aml system to run the pass and evaluation on GPU.~~

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Format your code by running `pre-commit run --all-files`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
Add `TorchTRTConversion` pass.

### CODE REVIEW

File: olive/passes/pytorch/trt_utils.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+def compile_trt_model(torch_module: torch.nn.Module, hidden_states: torch.Tensor, batch_size: int, seqlen: int):
+    """Compile a torch module to a TensorRT module.
+
+    :param torch_module: The torch module to compile. Only torch.nn.Linear modules are supported currently.
+    :param hidden_states: The input tensor to the torch module.
+    :param batch_size: The batch size of the input tensor.
+    :param seqlen: The maximum sequence length of the input tensor. seqlen dimension is treated as dynamic.
+    """
+
+    logging.getLogger("torch_tensorrt").setLevel(logging.ERROR)
```
Comment (guotuofeng) [inline olive/passes/pytorch/trt_utils.py:35]: what's the consideration that we need set the logger level of torch_tensor? would you add some comments here?

File: olive/passes/pytorch/torch_trt_conversion.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        # disable cache
+        use_cache = pytorch_model.config.use_cache
+        pytorch_model.config.use_cache = False
+
+        # get module list of layers
+        layers = get_layers(pytorch_model, model_type)
+
+        # get layer information
+        min_layer = config["min_layer"] or 0
+        max_layer = config["max_layer"] or len(layers)
```
Comment (guotuofeng) [inline olive/passes/pytorch/torch_trt_conversion.py:106]: do we need verify whether min_layer and max_layer is valid or not? what if user specify max_layer greater than len(layers)?

File: olive/passes/pytorch/trt_utils.py (hunk 1 above)
Comment (jambayk) [reply olive/passes/pytorch/trt_utils.py:35]: Yes, I can add comments. 
the torch tensorflow logs are very verbose so the olive logs get over populated. 
It produces many lines per module and there are around or so module per model.

File: olive/passes/pytorch/torch_trt_conversion.py (hunk 2 above)
Comment (jambayk) [reply olive/passes/pytorch/torch_trt_conversion.py:106]: Oh yes, I was thinking of the same but didn’t get around to adding this check. Also needed for the sparsegpt pass 

I will add it. Thanks!

**Comment by guotuofeng** [comment]
I think you might mean integration test to test TRTConversion. Is it possible that mock compile_trt_model to test the logic in the TorchTRTConversion in unit test?

**Comment by jambayk** [comment]
> I think you might mean integration test to test TRTConversion. Is it possible that mock compile_trt_model to test the logic in the TorchTRTConversion in unit test?

I wasn’t sure if it would fall under integration test if we use aml system + pass.
Yes, I can mock the function. That sounds like a good idea since that’s the only one that requires gpu. I can probably mock the cuda related code too.

File: olive/passes/pytorch/trt_utils.py (hunk 1 above)
Comment (jambayk) [reply olive/passes/pytorch/trt_utils.py:38]: Done!

File: olive/passes/pytorch/torch_trt_conversion.py (hunk 2 above)
Comment (jambayk) [reply olive/passes/pytorch/torch_trt_conversion.py]: Done!

File: test/unit_test/passes/pytorch/test_torch_trt_conversion.py
Comment (jambayk) [inline test/unit_test/passes/pytorch/test_torch_trt_conversion.py:1]: @guotuofeng I have added the ut. Can you please take a look at this? Thanks!

File: test/unit_test/passes/pytorch/test_torch_trt_conversion.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+@patch("torch.zeros", side_effect=mocked_torch_zeros)
+def test_sparsegpt(mock_is_cuda_available, mock_tensor_data_to_device, mock_module_to, mock_torch_zeros):
+    # setup
+    # mock trt utils since we don't have tensorrt and torch-tensorrt installed
+    mock_trt_utils = MagicMock()
+    mock_compile_trt_model = MagicMock()
+    mock_compile_trt_model.return_value = MockTRTLinearLayer()
+    mock_trt_utils.compile_trt_model = mock_compile_trt_model
+
+    sys.modules["olive.passes.pytorch.trt_utils"] = mock_trt_utils
```
Comment (guotuofeng) [inline test/unit_test/passes/pytorch/test_torch_trt_conversion.py:47]: >sys.modules["olive.passes.pytorch.trt_utils"] = mock_trt_utils [](http://example.com/codeflow?start=4&length=62)

I suppose this line is to avoid import TensorRT related stuff in UT. right?

Comment (jambayk) [reply test/unit_test/passes/pytorch/test_torch_trt_conversion.py:47]: yes, tenssort requires gpu, etc

File: test/unit_test/passes/pytorch/test_torch_trt_conversion.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        data_config = huggingface_data_config_template(model_name=model_name, task=task, **dataset)
+        config = {
+            "data_config": data_config,
+        }
+
+        p = create_pass_from_dict(TorchTRTConversion, config, disable_search=True)
+        output_folder = str(Path(tempdir) / "sparse")
+
+        # execute
+        model = local_system.run_pass(p, input_model, None, output_folder)
```
Comment (guotuofeng) [inline test/unit_test/passes/pytorch/test_torch_trt_conversion.py:76]: >local_system.run_pass [](http://example.com/codeflow?start=16&length=21)

could the pass.run be more clear?

Comment (jambayk) [reply test/unit_test/passes/pytorch/test_torch_trt_conversion.py:76]: The other pass tests do the same as this so I copied them here

File: test/unit_test/passes/pytorch/test_torch_trt_conversion.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+# all the following patches are needed to run the test on CPU
+# make torch.cuda.is_available return True
+@patch("torch.cuda.is_available", return_value=True)
+# make tensor_data_to_device return the input tensor, ignore device
+@patch("olive.common.utils.tensor_data_to_device", side_effect=lambda x, y: x)
+# make torch.nn.Module.to do nothing
+@patch("torch.nn.Module.to", side_effect=lambda device: None)
+# replace device in kwargs with "cpu"
+@patch("torch.zeros", side_effect=mocked_torch_zeros)
+def test_sparsegpt(mock_is_cuda_available, mock_tensor_data_to_device, mock_module_to, mock_torch_zeros):
```
Comment (guotuofeng) [inline test/unit_test/passes/pytorch/test_torch_trt_conversion.py:39]: >test_sparsegpt [](http://example.com/codeflow?start=4&length=14)

test_sparsegpt_happy_path be more clear since we might add other cases in future to test the invalid scenarios like max_layer invalid?

Comment (jambayk) [reply test/unit_test/passes/pytorch/test_torch_trt_conversion.py:39]: Yeah, I can do that

File: olive/passes/pytorch/torch_trt_conversion.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            del info["handles"]
+
+        # convert submodules to trt modules
+        logger.debug(f"Converting layers {min_layer} to {max_layer}...")
+        for layer_index, info in layer_info.items():
+            logger.debug(f"Converting layer {layer_index}...")
+            for name, shape in info["input_shapes"].items():
+                input = torch.zeros(shape, dtype=torch.float16, device=device)
+                # create trt module
+                trt_module = compile_trt_model(info["submodules"][name], input, batch_size, seqlen)
```
Comment (guotuofeng) [inline olive/passes/pytorch/torch_trt_conversion.py:153]: >compile_trt_model [](http://example.com/codeflow?start=29&length=17)

should we check whether the input submodule is torch.nn.Linear before calling compile_trt_model? or the pass caller need make sure it?

Comment (jambayk) [reply olive/passes/pytorch/torch_trt_conversion.py:153]: This is guaranteed by line 123. 
all of the sub modules we access here are found by that line where we specifically ask for NN.linear

Comment (guotuofeng) [reply olive/passes/pytorch/torch_trt_conversion.py:153]: If user need to make sure the parent model, should we add document for what kinds of model types supports?

Comment (guotuofeng) [reply olive/passes/pytorch/torch_trt_conversion.py:153]: or this pass only used for sparsegpt scenario?

Comment (jambayk) [reply olive/passes/pytorch/torch_trt_conversion.py:153]: Yes, we are only using it for the sparsegpt scenario. I can add documentation in the pass docstring

File: olive/passes/pytorch/torch_trt_conversion.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    get_layer_submodules,
+    get_layers,
+    seqlens,
+    validate_min_max_layers,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class TorchTRTConversion(Pass):
```
Comment (guotuofeng) [inline olive/passes/pytorch/torch_trt_conversion.py:27]: >TorchTRTConversion [](http://example.com/codeflow?start=6&length=18)

does this pass support any type PT model of just some models defined in the layer map? Do we need add document to demonstrate what type of models this pass support?

File: test/unit_test/passes/pytorch/test_torch_trt_conversion.py (hunk 4 above)
Comment (jambayk) [reply test/unit_test/passes/pytorch/test_torch_trt_conversion.py:76]: Resolving this conversation since this is the case for all pass tests. Maybe we can make the change to all pass tests in a different PR if nedded. 
Please unresolve the conversation if you think otherwise.

File: olive/passes/pytorch/torch_trt_conversion.py (hunk 7 above)
Comment (jambayk) [reply olive/passes/pytorch/torch_trt_conversion.py:27]: Done. Added to the docs and docstring. 
I have plans to support non hf_config pytorch models in the future. But for now, it only supports PyTorch model with hf_config and some model types.

**Review by guotuofeng — APPROVED** [review]
